Implement PTY support

patrickbkr opened the pull request:
This is not done yet. I still need to write tests and clean up the Git history and probably more fixes. But I want to push this out now to make people aware that I'm working on this and to receive feedback.

patrickbkr committed:
Implement PTY support

patrickbkr commented:
I've run into a design issue.
The PTY APIs on Unix and Windows differ. On Unix one receives one single bi-directional FD on the parent process side. On Windows one receives two unidirectional HANDLES on the parent process side.
Currently I give the caller of `uv_spawn` two pipes. One for `stdio[0]` and one for `stdio[1]`. On Windows they are bound to the two parent process HANDLES. On Unix both are bound to the same bi-directional FD. This works, but poses a problem on Unix, because when closing one of the two pipes the other one will also not be usable anymore.
Any ideas on how best to solve this?
@bnoordhuis

bnoordhuis commented:
> [..] works, but poses a problem on Unix, because when closing one of the two pipes the other one will also not be usable anymore

Just document that closing one renders the other one unusable / that you need to keep both open?

vtjnash commented:
I am somewhat opposed to adding new flags to uv_spawn, if we could instead do this with a separate API (similar to how uv_pipe and uv_socketpair are less buggy versions of the corresponding uv_spawn options), I would find that preferable

patrickbkr commented:
One more question:
On Windows, the PTY pipes need to be synchronous (that's a requirement of ConPTY). Thus I think I'll have to put writing and reading from the PTY pipes on separate threads to keep it non-blocking. Should I either:
- Just call `uv_queue_work`, this will make use of the libuv thread pool.
- Setup and manage two threads manually. This is more code, but doesn't touch the libuv thread pool.

bnoordhuis commented:
uv_queue_work is probably easier for now but do it in a way that leaves open the option of switching to dedicated threads in the future (e.g. don't embed the `uv_work_t` directly because we can't remove that without breaking ABI.)

patrickbkr commented:
Just a quick update. I believe to finally have Windows working. Need to test / fix MacOS and BSD. And a few minor nits here and there.

patrickbkr commented:
It's finally done. I consider this PR ready for review.

patrickbkr commented:
fixes https://github.com/libuv/libuv/issues/2640

patrickbkr commented on src/win/process.c:
```diff
@@ -814,6 +826,35 @@ static void CALLBACK exit_wait_callback(void* data, BOOLEAN didTimeout) {
+} uv__pty_close_pty_work;
+
+void pty_close_pty_cb(uv_work_t *req) {
+  HANDLE hLibrary = LoadLibraryExW(L"kernel32.dll", 0, 0);
```
Thanks for having a look!

I actually don't know first hand whether including an external library dll in libuv is actually a no-go. I just strongly assume so.
But maybe there could be a compile time opt-in to linking against an externally provided conpty dll. It would then require building libuv oneself to be able to do so.

@bnoordhuis Would such a compile time opt-in be acceptable?

@Tyriar Would that suffice for your use case? Given you have experience with doing the *Windows provided ConPTY XOR custom ConPTY DLL dance*, could you imagine trying your hand on adding such support to this PR?

Tyriar replied:
I'm a TS developer at heart these days and haven't done much C++ for a long time, so I'm not so confident in contributing this myself.

I'm also not sure a compile time check would be sufficient for our case as that would mean being forced to build a custom Electron and standalone node in order to enable this support, since we use node-pty in both. A string to a DLL like node-pty has may be atypical and maybe have security considerations, but that would make it convenient to consume.

patrickbkr replied:
> A string to a DLL like node-pty has may be atypical and maybe have security considerations, but that would make it convenient to consume.

Just so I understand correctly, in node-pty it is possible to pass (at run time?) a string which makes node-pty use the DLL at the respective path? Do you have an idea how something like that could look for libuv? How would one pass that string? Via a separate function call setting a global variable? As another field in `uv_process_options_t`?

@bnoordhuis What's your opinion on this?

bnoordhuis replied:
> Would such a compile time opt-in be acceptable?

No, but a runtime opt-in might be. `uv_process_options_t` cannot be changed due to ABI compatibility but a new `uv_process_flags` could be introduced that makes libuv switch to an extended definition of that struct, one with space for additional fields.

(That's strictly speaking an aliasing violation but that's why libuv is, and consumers should be, compiled with `-fno-strict-aliasing`.)

patrickbkr replied:
@bnoordhuis Intuitively I'd have opted to create a separate function `uv_set_conpty_dll` that sets a global, similar to `uv_replace_allocator`. I can't imagine a situation where I'd want to use different ConPTY modes for different processes I'm spawning. Seems to be easier to use and less fiddly than juggling differently sized `uv_process_options_t`. Do you agree?

vtjnash replied:
This goes back to why I said this is not a good API design to mix concerns like this, and should instead be implemented like uv_pipe is separate from uv_spawn. You may need to introduce a pseudo handle kind in libuv, which is just an API for dealing with this specific handle, but it avoids entirely the configuration problems by not making it expected this one API needs to shoehorn in all functionality (https://github.com/libuv/libuv/pull/4802#issuecomment-2973439694)

deepak1556 replied:
> Just so I understand correctly, in node-pty it is possible to pass (at run time?) a string which makes node-pty use the DLL at the respective path?

Some more context, the addon does not support the option to accept path to conpty.dll instead the native addon ships the conpty.dll alongside so a boolean api was sufficient to flip how the loadlibrary calls should look like within the addon https://github.com/microsoft/node-pty/blob/24b53bbf573ed7ab35f8897aa8b690dbd9064fc9/src/win/conpty.cc#L164-L189

For libuv, it would be the embedders shipping the conpty.dll. So a runtime api suggested by @bnoordhuis when configuring the event loop would help, the embedder could then resurface it as some public api very likely a per process api.

patrickbkr replied:
I start to think that a plain string path pointing to a DLL file is a can of worms. What about relative paths? What about path length?

Am I overreacting? Are there better alternatives?

patrickbkr commented:
Irrespective of the above discussion about adding support for a user supplied ConPTY DLL, this PR is still in good shape. I believe user supplied ConPTY DLL support can easily be added in a separate PR.

From all I see the only feedback concerning this PR itself are the concerns vtjnash brought up.
@vtjnash Did you notice [my ask](https://github.com/libuv/libuv/pull/4802#discussion_r2314951740) for more explanation of your concerns?

bnoordhuis commented:
Speaking for myself, I'm fine with leaving out DLL loading for now. I agree it's something that can always be added later if needed.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -379,6 +382,20 @@ static void uv__process_child_init(const uv_process_options_t* options,
+    // and process group leader.
+    if (setsid() < 0) {
+        uv__write_errno(error_fd);
+    }
```
Style issues: C89-style comments and no braces around single line `if` statements.
```suggestion
    /* Put ourself into a new session and process group, making us session
      * and process group leader.
      */
    if (setsid() < 0)
        uv__write_errno(error_fd);
```

bnoordhuis commented on src/unix/process.c:
```diff
@@ -963,6 +983,74 @@ static int uv__spawn_and_init_child(
+    }
+
+    int path_tty_size = 40;
+    char *path_tty = uv__malloc(path_tty_size * sizeof(char));
```
The realloc loop isn't really necessary, you can declare a `char path_tty[TTY_NAME_MAX + 1] = {0};` buffer on the stack.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -1009,18 +1112,33 @@ int uv_spawn(uv_loop_t* loop,
   }
 
+  if (options->flags & UV_PROCESS_PTY) {
+    if ((err = uv__spawn_make_pty(&process->pty_fd, &fd_tty, options->pty_cols, options->pty_rows)) != 0)
```
Please keep lines <= 80 columns.

bnoordhuis reviewed (commented):
Only a partial review. See especially my comment about API/ABI stability. I can't merge this as-is because of that.

patrickbkr committed:
Address a few review comments

patrickbkr committed:
Add a BSD implementation

patrickbkr committed:
Better error handling and docs re non-blocking PTY on unix

patrickbkr committed:
ABI compat: Don't change the size of uv_process_t

Use the u.fd space conveniently added in UV_HANDLE_FIELDS.

patrickbkr committed:
ABI compat: Don't change the size of uv_process_options_t

Do so by introducing a `uv_spawn2` and `uv_process_options2_t`.

patrickbkr committed:
Fix MacOS build

The TTY_NAME_MAX constant isn't available on MacOS. Use
`sysconf(_SC_TTY_NAME_MAX)` instead.

patrickbkr committed:
Fix a "unused function" build warning

lhecker commented on src/win/process.c:
```diff
@@ -886,17 +932,208 @@ void uv__process_endgame(uv_loop_t* loop, uv_process_t* handle) {
+  if (!hLibrary)
+    return uv_translate_sys_error(GetLastError());
+
+  PFNRESIZEPSEUDOCONSOLE pfnResize = (PFNRESIZEPSEUDOCONSOLE)GetProcAddress((HMODULE)hLibrary,"ResizePseudoConsole");
```
It would be nice if these could get lazy initialized as global variables. `LoadLibrary` and `GetProcAddress` are both fairly expensive calls. An SRWLOCK may be suitable for this. I would also suggest passing `LOAD_LIBRARY_SEARCH_SYSTEM32`, but for `kernel32.dll` in particular I'd just use `GetModuleHandleW`. It won't be getting unloaded.

lhecker reviewed (commented):
(Great work!)

patrickbkr committed:
Persist and lazily initialize `ResizePseudoConsole` func

Both GetProcAddress as well as LoadLibrary are fairly expensive. So cache
the result. This is racy, but it doesn't matter much, on a collision we'll
do a bit of throw away work, but the assignment of `pfnResize` is atomic.
So nothing bad will happen.

patrickbkr committed:
Enlarge the PTY out read buffer

Should be at least 128KiB according to lhecker++.

patrickbkr committed:
Speed up `uv_pty_resize` even more

kernel32.dll will always be loaded, so we can just retrieve its handle
instead of loading it.

patrickbkr commented:
@lhecker Thanks for your review! Changes are coming up.

patrickbkr committed:
Make PTY tests more robust

Let's just punt on the EOF vs EIO issue.

patrickbkr commented:
I've incorporated the feedback of lhecker++ and did a rebase on latest v1.x. Now again ready for review.

bnoordhuis commented:
@patrickbkr just wanted to let you know I haven't forgotten about this PR, it's first on my list when I have a longer stretch of review time.

lygstate commented:
> @patrickbkr just wanted to let you know I haven't forgotten about this PR, it's first on my list when I have a longer stretch of review time.

ping for review. @bnoordhuis
